### ISSUE TITLE
Send flatcc buf with message queue

### ISSUE THREAD

**Issue by FerenczD** [dvidelabs/flatcc#107]
My problem is that I'm trying to send a buf through message queues on linux. So I finalize the buffer with

`buf = flatcc_builder_get_direct_buffer(&builder, &size);`

and send it after with 

`mq_send(THREAD2_MQ, buf, sizeof(buf), 0); `

On the receiving end I have a char* buffer that is just getting nulls

I understand that buf is a void* and you can't send pointers, but in your example you say you are able to send the buffer over the network and that buf is an snapshot of the builder. 

Am I missing a function or casting the buf into something I can send with message queue? 

Side note: I tried casting buf to (char*) as mq_send() specifies.
I also tried receiving the buf as a void* and was still giving null

**Comment by mikkelfj** [comment]
`sizeof(buf)` will usually, but not always, give you the size of the pointer and not the underlying buffer. If this is not just a mistake and you need to understand this better, the book "C Programming Language, 2nd Edition 2nd Edition" by Kernighan and Ritchie is the one to get.

In this case you need the `size` variable` which will be set as a result of get direct buffer.
Also, if your buffer is large, then buf and size will be 0 and you need a more expensive call. But for short messages this is fine (ca 2K).

I have not worked with `mq_send`, so I can't say what is right, but I can say that you need to use the correct size:

    mq_send(THREAD2_MQ, buf, size, 0);

**Comment by derosier** [comment]
You've not given the full code, but I'm assuming you've declared buf as:
char * buf;
OR
void * buf;

sizeof(buf) would then equal 4 or 8 depending on the architecture of
your machine. Hence you're likely only sending the first word of the
buffer, which is probably all zeros, hence your result.

Based on this and other questions you've asked, I would guess you're
new to C?  May I suggest "C Programming Language, 2nd Edition" by
Kernighan and Ritchie: https://amzn.to/2IgoULi
It's considered the standard text on C programming.

And based on the syscalls you're using, if you're doing serious system
programming, you'll also want "The Linux Programming Interface: A
Linux and UNIX System Programming Handbook 1st Edition" by Michael
Kerrisk:
https://amzn.to/2GNc7gT

- Steve

**Comment by mikkelfj** [comment]
@derosier I guess we have the same taste in books :)

**Comment by FerenczD** [comment]
Thanks for the help. You were right, with the size I was sending was wrong, that was one of my errors. I just had to change the call to `size `instead of `sizeof(buf). `

its funny, I tried using `size `before but because it was also giving me errors I assumed that was not the reason :) my second error which I didn't think about was that I was calling a deserialize function I made which would take a` void*` instead of `char*` which is why it was not deserializing correctly and giving me nulls. 

Anyhow, it is working now. Appreciate the help.
